pyomo/core/expr/logical_expr.py

```diff
@@ -240,6 +239,26 @@ def atleast(n, *args):
     return result
 
 
+def all_different(*args):
+    """Creates a new AllDifferentExpression
+
+    Requires all of the arguments to take on a different value
+
+    Usage: all_different(m.X1, m.X2, ...)
+    """
+    return AllDifferentExpression(list(_flattened(args)))
```

Comment:
What verifies that all arguments are NumericExpressions?

Reply:
I just added a numeric and Boolean version of `_flattened` that will check for these and the other logical expressions.